File: 76_Minimum_Window_Substring/solution.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        /*
            Input: Strings s and t
            Output: string minimum window substring of s such that every character in t (all counts) is in the window. else return ""

            Strategy:
            1.) Sliding window
                - one pass on string t and store the character counts in a hashmap
                - while r < s.length
                    - if right character exist in t 
                        - if count > 0 (we do not want to reduce when we encounter repeats that are not in t)
                            - decrement count
                        - decrement map count
                    expand window
                    - when count reaches 0, we found a potential match where all chars are in window
                        - shortlist min length -> update global min length and start and end pointer
                        - if left char in t
                            - increment map count 
                            - increment counter if map count > 0 (got rid of repeats not included in t)
                        shrink window
                
                if min length is still INT_MAX, return "" else s.substr(start, end-start+1)

        */

        int l = 0, r = 0, count = 0, start = 0, end = 0;
        int min_length = INT_MAX;
        std::unordered_map<char,int>t_count;
        for (char c:t){++t_count[c];++count;}

        while (r<s.length()){
            if (t_count.find(s[r])!= t_count.end()){
                if (t_count[s[r]] > 0){
                    --count;
                }
                --t_count[s[r]];
            }
            ++r;

            while (count == 0){
                if (r-l < min_length){
                    start = l;
                    end = r;
                    min_length = r-l;
                }
                if (t_count.find(s[l]) != t_count.end()) {
                    ++t_count[s[l]];
                    if (t_count[s[l]] > 0){
                        ++count;
                    }
                }
                ++l;
            }
        }

        if (min_length == INT_MAX){
            return "";
        }
        return s.substr(start,min_length);
        
    }
};

File: 76_Minimum_Window_Substring/solution.java
class Solution {
    public boolean compareMap(HashMap<Character, Integer> map1, HashMap<Character, Integer>map2) {
        if (!map1.keySet().equals(map2.keySet())) return false;
        for (char ch: map1.keySet()) {
            if (map1.get(ch) < map2.get(ch)) return false;
        }
        
        return true;
    }
    public String minWindow(String s, String t) {
        /*
            Input: String s, String t
            Output: String the minimum window substring, every character in t is included in the window.
            If there is no such substring, return the empty string "".

            Sample test cases
            1.) s = "ADOBECODEBANC", t = "ABC", output = "BANC"
                - "ADOBEC" - length 6
                - "BECODEBA" - length 8
                - "CODEBA" - length 6
                - "BANC" - length 4

            Strategy
            - Form the different substrings where every character in t (including) duplicates is included.
            - Since t may contain duplicates, we will need a hashmap to know if all the characters and 
            counts have been found.
            - while there exist characters with count < what is contained in t, expand the window
            - while all characters in the existing window >= what is contained in t,
                - shortlist length
                - reduce window size and remove character count
            - process continues as long as right end of window < length of s

            Alternatively
            - Fill up the character counts for characters present in t that are in s; 
                - begin with left: 0, right: s.length-1;
            - while character counts are >= those present in t, update length and reduce window size
        */
        HashMap<Character, Integer> s_map = new HashMap<Character, Integer>();
        HashMap<Character, Integer> t_map = new HashMap<Character, Integer>();
        int count;
        int left = 0, right = 0;
        int n = s.length();

        String res = "";
        
        for (char ch: t.toCharArray()) {
            count = t_map.getOrDefault(ch,0);
            t_map.put(ch, count+1);
        }

        while (right < n) {
            while (right < n && !compareMap(s_map, t_map)) {
                if (t_map.containsKey(s.charAt(right))){
                    count = s_map.getOrDefault(s.charAt(right),0);
                    s_map.put(s.charAt(right),count+1);
                }
                ++right;
            }
            
            while (left < right & compareMap(s_map, t_map)) {
                if (res.length() == 0 || res.length() > right-left) res = s.substring(left,right);
                
                if (s_map.containsKey(s.charAt(left))) {
                    count = s_map.get(s.charAt(left));
                    s_map.put(s.charAt(left),count-1);
                    
                }
                ++left;
                
            }
            
        }
        return res;


    }
}
